File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Golang/Carro.go
package main

import "fmt"

type Motor struct {
	tipo     string
	potencia int
	ligado   bool
}

func (m *Motor) ligar() {
	m.ligado = true
	fmt.Println("O Motor está ligado")
}

func (m *Motor) desligar() {
	m.ligado = false
	fmt.Println("O Motor está desligado")
}

type Roda struct {
	marca   string
	tamanho int
	inflado bool
}

func (r *Roda) inflar() {
	r.inflado = true
	fmt.Println("A Roda está inflada")
}

func (r *Roda) desinflar() {
	r.inflado = false
	fmt.Println("A Roda está desinflada")
}

type Carro struct {
	Motor
	Rodas      []Roda
	marca      string
	modelo     string
	ano        int
	ligado     bool
	velocidade float64
}

func (c *Carro) exibirDetalhes() {
	fmt.Printf(`
		Exibindo detalhes do Carro
		Motor {Tipo: %s, Potencia: %d}
		Rodas {Marca: %s, Tamanho: %d}
		Carro {Marca: %s, Modelo: %s, Ano: %d}
	`, c.tipo, c.potencia, c.Rodas[0].marca, c.Rodas[0].tamanho, c.marca, c.modelo, c.ano)
}

func (c *Carro) ligarCarro() int {
	if !c.Motor.ligado {
		fmt.Println("Não é possível ligar o carro, o motor está desligado")
		return 0
	}

	for _, r := range c.Rodas {
		if !r.inflado {
			fmt.Println("Não é possível ligar o carro, tem pelo menos uma roda desinflada")
			return 0
		}
	}

	c.ligado = true
	fmt.Println("O Carro está ligado")
	return 1
}

func (c *Carro) acelerar(v float64) {
	if c.ligado {
		fmt.Println("\nAcelerando carro")
		c.velocidade += v
		fmt.Printf("\nO Carro está a %.2f km/h", c.velocidade)
	} else {
		fmt.Println("\nO Carro está desligado")
	}
}

func (c *Carro) frear(v float64) {
	if c.ligado {
		fmt.Println("\nFreando carro")
		c.velocidade -= v
		if c.velocidade <= 0 {
			c.velocidade = 0
			fmt.Println("\nO Carro parou")
		} else {
			fmt.Printf("\nO Carro está a %.2f km/h", c.velocidade)
		}
	} else {
		fmt.Println("\nO Carro está desligado")
	}
}

File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Python/Carro.py
class Motor:
    def __init__(self, tipo, potencia):
        self.tipo = tipo
        self.potencia = potencia
        self.ligado = False

    def ligar(self):
        self.ligado = True
        print("O motor está ligado.")

    def desligar(self):
        self.ligado = False
        print("O motor está desligado.")
        
class Pneu:
    def __init__(self, marca, tamanho):
        self.marca = marca
        self.tamanho = tamanho
        self.inflado = False

    def inflar(self):
        self.inflado = True
        print("O Pneu está inflado.")

    def desinflar(self):
        self.inflado = False
        print("O Pneu está desinflado.")

class Carro:
    def __init__(self, Motor, Pneus, ano, marca, modelo):
        self.Motor = Motor
        self.Pneus = Pneus
        self.ano = ano
        self.marca = marca
        self.modelo = modelo
        self.ligado = False
        self.velocidade = 0
 
    def ligar(self):
        self.Motor.ligar()
        self.ligado = True
        print("O carro está pronto para rodar.")

    def desligar(self):
        self.Motor.desligar()
        self.ligado = False
        print("O carro está desligado.")

    def trocar_pneu(self, indice, novo_pneu):
        pneu_removido = self.pneu[indice]
        self.pneu[indice] = novo_pneu
        print(f"Pneu {indice + 1} foi trocado")
        return pneu_removido
    
    def exibirDetalhes(self):
        print(f"""\tExibindo detalhes do carro 
              \tMotor [Tipo: {self.Motor.tipo}; Potencia: {self.Motor.potencia}]
              \tPneus [Marca: {self.Pneus[0].marca}; Tamanho: {self.Pneus[0].tamanho}]
              \tCarro [Marca: {self.marca}; Modelo: {self.modelo}; Ano: {self.ano}]
        """)

    def acelerar(self, vel):
        if (self.ligado):
            print("Acelerando carro")
            self.velocidade += vel
            print(f"O Carro está a {self.velocidade} km/h")
        else:
            print("O Carro está desligado")

    def frear(self, vel):
        if (self.ligado):
            print("Freando carro")
            self.velocidade -= vel
            if self.velocidade <= 0:
                self.velocidade = 0
            print(f"O Carro está a {self.velocidade} km/h")
        else:
            print("O Carro está desligado")
